File: backend/app/schemas/product.py
from datetime import datetime
from pydantic import BaseModel, ConfigDict, Field

from app.schemas.category import CategoryOut


class ProductOut(BaseModel):
    id: int
    name: str
    description: str | None = None
    price: float
    is_active: bool
    created_at: datetime
    updated_at: datetime
    categories: list[CategoryOut] = []

    model_config = ConfigDict(from_attributes=True)


class ProductCreate(BaseModel):
    name: str = Field(min_length=1, max_length=150)
    description: str | None = Field(default=None, max_length=500)
    price: float = Field(gt=0)
    is_active: bool = True


class ProductUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=150)
    description: str | None = Field(default=None, max_length=500)
    price: float | None = Field(default=None, gt=0)
    is_active: bool | None = None

File: frontend/types/product.ts
// Mirrors: app/schemas/product.py, app/schemas/product_variant.py, app/schemas/category.py, app/schemas/product_category.py

export interface Category {
  id: number;
  name: string;
  description: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface CategoryCreate {
  name: string;
  description?: string | null;
}

export interface CategoryUpdate {
  name?: string;
  description?: string | null;
}

export interface Product {
  id: number;
  name: string;
  description: string | null;
  price: number;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  categories: Category[];
}

export interface ProductCreate {
  name: string;
  description?: string | null;
  price: number;
  isActive?: boolean;
}

export interface ProductUpdate {
  name?: string;
  description?: string | null;
  price?: number;
  isActive?: boolean;
}

export interface ProductVariant {
  id: number;
  productId: number;
  sku: string;
  price: number;
  stockQuantity: number;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface ProductVariantCreate {
  sku: string;
  price: number;
  stockQuantity?: number;
  isActive?: boolean;
}

export interface ProductVariantUpdate {
  sku?: string;
  price?: number;
  stockQuantity?: number;
  isActive?: boolean;
}

export interface ProductCategory {
  id: number;
  productId: number;
  categoryId: number;
  category: Category;
  createdAt: string;
  updatedAt: string;
}

export interface ProductCategoryCreate {
  categoryId: number;
}
